### ISSUE TITLE
dbGetQuery() marks character strings with wrong encoding (UTF-8)

### ISSUE THREAD

**Issue by huftis** [rstats-db/odbc#67]
### Issue Description and Expected Result
`dbGetQuery()` marks character strings with wrong encoding (UTF-8).

The actual strings are encoded as ISO-8859-1 (latin1) both in the database and as returned by `dbGetQuery()`, but they are *marked* as having an UTF-8 encoding in R.

The expected result is that character columns are marked as having the correct encoding, i.e. the character encoding they are actually encoded in.

### Database
From dbGetInfo():
```
$dbms.name
[1] "Microsoft SQL Server"

$db.version
[1] "11.00.6020"
```

### Example
This is from a non-public database, so not reproducible, but basically I do:

```r
res = dbGetQuery(db, "select FstName from Person")
```

If the Person table has one row, with `Fstname` equal to `Øystein`, the result is displayed as:

```r
     FstName
1 <d8>ystein
```

and `res$FstName` is displayed as `[1] "\xd8ystein"`, indicating that the string is actually a `latin1` string. However, it is marked as having encoding UTF-8 in R, i.e. `Encoding(res$FstName)` returns `UTF-8`.

Changing the encoding of the returned object fixes the problem:

```r
> Encoding(res$FstName) = "latin1"
> res
  FstName
1 Øystein
```

(As the database encoding is the same as the native encoding, setting the `Encoding()` to `"unknown"` instead of `"latin1"` *also* works.)

The data is stored as a `latin1` string in the database, as shown by:

```r
> dbGetQuery(roas_db, "select CHARACTER_SET_NAME from information_schema.columns where COLUMN_NAME='FstName' and TABLE_NAME = 'Person'")
  CHARACTER_SET_NAME
1              iso_1
```

### Suggested fix

Either:
  1. Reencode strings coming from the database to UTF-8, and keep the `Encoding()` in R as UTF-8.
  2. Mark the strings with the encoding from the databse, e.g. with `Encoding(colname) = "latin1"` for variables with the `iso_1` encoding.

<details>
<summary>Session Info</summary>

```r
> devtools::session_info()
Session info ------------------------------------------------------------------------------------------------------------------------------------------
 setting  value                        
 version  R version 3.4.0 (2017-04-21) 
 system   x86_64, mingw32              
 ui       RStudio (1.0.143)            
 language (EN)                         
 collate  Norwegian-Nynorsk_Norway.1252
 tz       Europe/Berlin                
 date     2017-06-07                   

Packages ----------------------------------------------------------------------------------------------------------------------------------------------
 package   * version date       source        
 base      * 3.4.0   2017-04-21 local         
 blob        1.0.0   2016-12-29 CRAN (R 3.4.0)
 compiler    3.4.0   2017-04-21 local         
 datasets  * 3.4.0   2017-04-21 local         
 DBI       * 0.6-1   2017-04-01 CRAN (R 3.4.0)
 devtools    1.13.0  2017-05-08 CRAN (R 3.4.0)
 digest      0.6.12  2017-01-27 CRAN (R 3.4.0)
 graphics  * 3.4.0   2017-04-21 local         
 grDevices * 3.4.0   2017-04-21 local         
 memoise     1.1.0   2017-04-21 CRAN (R 3.4.0)
 methods   * 3.4.0   2017-04-21 local         
 odbc      * 1.0.1   2017-02-07 CRAN (R 3.4.0)
 Rcpp        0.12.10 2017-03-19 CRAN (R 3.4.0)
 stats     * 3.4.0   2017-04-21 local         
 tibble      1.3.0   2017-04-01 CRAN (R 3.4.0)
 tools       3.4.0   2017-04-21 local         
 utils     * 3.4.0   2017-04-21 local         
 withr       1.0.2   2016-06-20 CRAN (R 3.4.0)
```
</details>

**Comment by jimhester** [comment]
This should now be fixed in the latest master. You will need to use the argument `encoding = "latin1"` to `dbConnect()` when you are creating the connection, and strings should be properly decoded to UTF-8 when retrieved.

**Comment by carioca67** [comment]
Wow, Kirill I am AMAZED at your knowledge and efficienty. I did follow your link and found a solution: the dbConnect call should contain an extra "encoding" parameter (see below), thanks my friend!

con <- dbConnect(odbc::odbc(),
                   Driver    = "SQL Server", 
                   Server    = "sss",
                   Database  = "ddd",
                   UID       = "uuu",
                   PWD       = askForPassword("passwd: "),
                   Port      = 1433,
                   encoding = "latin1") # fixed in odbc

**Comment by kummerer94** [comment]
Thanks for this thread! It helped me resolve a similar issue we had when using a Microsoft SQL database and the ODBC Driver provided by Microsoft.

<details><summary>The `dbGetInfo()` output.</summary>
<p>

```$dbname
[1] "test"

$dbms.name
[1] "Microsoft SQL Server"

$db.version
[1] "12.00.0041"

$username
[1] "dbo"

$host
[1] ""

$port
[1] ""

$sourcename
[1] ""

$servername
[1] "xxx"

$drivername
[1] "msodbcsql18.dll"

$odbc.version
[1] "03.80.0000"

$driver.version
[1] "18.00.0001"

$odbcdriver.version
[1] "03.80"

$supports.transactions
[1] TRUE

$getdata.extensions.any_column
[1] FALSE

$getdata.extensions.any_order
[1] FALSE

attr(,"class")
[1] "Microsoft SQL Server" "driver_info"          "list"
```
</p>
</details>


However, I would like to add that unlike @jimhester has described, **I specifically had to leave the `encoding` in `dbConnect()` blank**. When I set `encoding = "latin1"`, the resulting rows in the database were broken. So, if anyone like me arrives here and wonders what they have to do:

When reading or writing any strings to the database, make sure to use `enc2utf8()` to encode your strings in UTF-8 before passing them to the DBI methods. This will do the trick - at least for my configuration.

**Comment by shrektan** [comment]
@kummerer94 If you find the resulting rows be broken when you set `encoding = "latin1"` and works with blank, your database probably is not encoded in "latin1" but the same with your current locale.

**Comment by kummerer94** [comment]
@shrektan Thank you, I really appreciate your help (since I am not really proficient in this topic...) but it appears as though the database columns are indeed encoded in `latin1` (I find that the encoding is set to `SQL_Latin1_General_CP1_CI_AS` - I assume there is no difference?). The database encoding is also `SQL_Latin1_General_CP1_CI_AS`.

I used:
```sql
SELECT DATABASEPROPERTYEX('test', 'Collation') SQLCollation;
```

and

```sql
SELECT t.name TableName, c.name ColumnName, collation_name  
FROM sys.columns c  
INNER JOIN sys.tables t on c.object_id = t.object_id where t.name = 'TABLENAME';
```

**Comment by shrektan** [comment]
May I ask you the OS you are working on? A Windows machine or a Linux machine? If your OS is Linux, then it explains the reason: The MS SQLServer Driver on Linux changes the encoding of the data to UTF-8 by default.

**Comment by kummerer94** [comment]
I am on Windows.

```
R version 4.1.2 (2021-11-01)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 22000)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252 LC_NUMERIC=C                           LC_TIME=English_United States.1252
system code page: 65001

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base

other attached packages:
[1] InternalPackageY_0.6.0 testthat_3.1.1    internalPackageX_1.3.3

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.8          lubridate_1.8.0     lattice_0.20-45     zoo_1.8-9           prettyunits_1.1.1   ps_1.6.0            assertthat_0.2.1    rprojroot_2.0.2     digest_0.6.29       utf8_1.2.2
[11] R6_2.5.1            odbc_1.3.3          httr_1.4.2          pillar_1.6.4        tictoc_1.0.1        rlang_0.4.12        rstudioapi_0.13     data.table_1.14.2   callr_3.7.0         blob_1.2.2
[21] R.utils_2.11.0      R.oo_1.24.0         drat_0.2.2          config_0.3.1        desc_1.4.0          devtools_2.4.3      AzureAuth_1.3.3     readr_2.1.1         bit_4.0.4           AzureKeyVault_1.0.5
[31] compiler_4.1.2      httpuv_1.6.5        pkgconfig_2.0.3     pkgbuild_1.3.1      AzureGraph_1.3.2    tidyselect_1.1.1    tibble_3.1.6        fansi_1.0.2         crayon_1.4.2        dplyr_1.0.7
[41] tzdb_0.2.0          dbplyr_2.1.1        withr_2.4.3         later_1.3.0         R.methodsS3_1.8.1   rappdirs_0.3.3      grid_4.1.2          jsonlite_1.7.3      lifecycle_1.0.1     DBI_1.1.2
[51] git2r_0.29.0        magrittr_2.0.1      cli_3.1.0           cachem_1.0.6        renv_0.15.1         fs_1.5.2            promises_1.2.0.1    remotes_2.4.2       ellipsis_0.3.2      keyring_1.3.0
[61] vctrs_0.3.8         generics_0.1.1      tools_4.1.2         bit64_4.0.5         glue_1.6.0          purrr_0.3.4         hms_1.1.1           processx_3.5.2      pkgload_1.2.4       fastmap_1.1.0
[71] yaml_2.2.1          sessioninfo_1.2.2   memoise_2.0.1       AzureRMR_2.4.3      usethis_2.1.5
```

**Comment by shrektan** [comment]
@kummerer94 

There's `system code page: 65001` below `locale`. I think you enables the UTF-8 stuff on Windows. I'm not sure about the consequences. 

Can you please attach the result of `l10n_info()` and `"\u4e2d\u6587" |> enc2native() |> charToRaw()`?

From the first we can know more about the specifics of your OS locale. From the second we can verify what your "real" native encoding is.


In addition, this is very werid, as `odbc` already convers everything to utf8 (https://github.com/r-dbi/odbc/commit/7c763d5fa572a71f72c4302b36f4e7a8affae41e)

> When reading or writing any strings to the database, make sure to use enc2utf8() to encode your strings in UTF-8 before passing them to the DBI methods. This will do the trick - at least for my configuration.

**Comment by kummerer94** [comment]
@shrektan 

My `l10n_info()` is the following:

```
r$> l10n_info()
$MBCS
[1] FALSE

$`UTF-8`
[1] FALSE

$`Latin-1`
[1] TRUE

$codepage
[1] 1252

$system.codepage
[1] 65001
```

Your command is evaluated to:

```
r$> "\u4e2d\u6587" |> enc2native() |> charToRaw()
 [1] 3c 55 2b 34 45 32 44 3e 3c 55 2b 36 35 38 37 3e
```

> In addition, this is very werid, as odbc already convers everything to utf8 (https://github.com/r-dbi/odbc/commit/7c763d5fa572a71f72c4302b36f4e7a8affae41e)

It is very weird and I have verified again that I am using the most recent release of odbc (1.3.3).

**Comment by jkroes** [comment]
This doesn't seem to be fixed in v1.3.3. Using `dbConnect(..., encoding="latin1") %>% collect` returns strings marked as UTF-8 but with improperly decoded(?) latin1 bytes. E.g., en hyphens (hex 96) show up as "\u0096" (hex c2 96). I was able to fix the strings with `iconv(..., "UTF-8", "latin1"), but

**Comment by diegomsg** [comment]
Hi. Had similar issues as @kummerer94
The result of `l10n_info()`
```
$MBCS
[1] FALSE

$`UTF-8`
[1] FALSE

$`Latin-1`
[1] TRUE

$codepage
[1] 1252

$system.codepage
[1] 1252
```
and the other code gave me the same result:
```
"\u4e2d\u6587" %>% enc2native() %>% charToRaw()
 [1] 3c 55 2b 34 45 32 44 3e 3c 55 2b 36 35 38 37 3e
```
SQL table encoding is `SQL_Latin1_General_CP1_CI_AS`
Testing different codepages gave me good results using she same codepage in encoding string in `dbConnect()`
```
dbConnect(odbc(),
  Driver = "SQL Server",
  Server = "XXX",
  Database = "XXX",
  Trusted_Connection = "True",
  encoding = "1252"
)
```
